rln/src/merkle_tree.rs

```diff
@@ -1,251 +1,231 @@
-const fn parent(index: usize) -> Option<usize> {
-    if index == 0 {
-        None
-    } else {
-        Some(((index + 1) >> 1) - 1)
-    }
-}
+pub struct MerkleProof<H: Hasher>(pub Vec<(H::Fr, u8)>);
 
-/// For a given node index, return index of the first (left) child.
-const fn first_child(index: usize) -> usize {
-    (index << 1) + 1
-}
+impl<H> MerkleTree<H>
+where
+    H: Hasher,
+{
+    pub fn new(depth: usize, default_leaf: H::Fr) -> Self {
+        let mut cached_nodes: Vec<H::Fr> = Vec::with_capacity(depth + 1);
+        cached_nodes.push(default_leaf);
+        for i in 0..depth {
+            cached_nodes.push(H::hash(&[cached_nodes[i]; 2]));
+        }
+        cached_nodes.reverse();
+        MerkleTree {
+            cached_nodes: cached_nodes.clone(),
+            depth,
+            nodes: HashMap::new(),
+            next_index: 0,
+        }
+    }
 
-const fn levels(index: usize) -> usize {
-    // `n.next_power_of_two()` will return `n` iff `n` is a power of two.
-    // The extra offset corrects this.
-    (index + 2).next_power_of_two().trailing_zeros() as usize - 1
-}
+    pub fn tree_size(&self) -> usize {
+        1 << self.depth
+    }
 
-impl<H: Hasher> MerkleTree<H> {
-    /// Creates a new `MerkleTree`
-    /// tree_height - the height of the tree made only of hash nodes. 2^tree_height is the maximum number of leaves hash nodes
-    pub fn new(tree_height: usize, initial_leaf: H::Hash) -> Self {
-        // total number of levels of the tree, i.e. # of layers including pre-images layer of tree leaves, thus equal to tree_height+1
-        let levels = tree_height + 1;
-        // Compute empty node values, leaf to root
-        let empty = successors(Some(initial_leaf), |prev| Some(H::hash_node(prev, prev)))
-            .take(levels)
-            .collect::<Vec<_>>();
-
-        // Compute node values
-        let nodes = empty
-            .iter()
-            .rev()
-            .enumerate()
-            .flat_map(|(levels, hash)| repeat(hash).take(1 << levels))
-            .cloned()
-            .collect::<Vec<_>>();
-        debug_assert!(nodes.len() == (1 << levels) - 1);
-
-        let next_index = 0;
-
-        Self {
-            levels,
-            empty,
-            nodes,
-            next_index,
-        }
+    pub fn root(&self) -> H::Fr {
+        self.get_node(0, 0)
     }
 
-    #[must_use]
-    pub fn num_leaves(&self) -> usize {
-        self.levels
-            .checked_sub(1)
-            .map(|n| 1 << n)
-            .unwrap_or_default()
+    pub fn set(&mut self, index: usize, leaf: H::Fr) -> io::Result<()> {
+        if index >= self.tree_size() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "index exceeds set size",
+            ));
+        }
+        self.nodes.insert((self.depth, index), leaf);
+        self.recalculate_from(index);
+        self.next_index = max(self.next_index, index + 1);
```

Comment:
Oh, that gets complicated, such option makes sense in general use-cases but not for the RLN membership Merkle tree, right? do you see a specific use-case for such sparse leaf assignment? it might be easier to remove this flexibility i.e., the `set` function 
Also, another question would be why `N` up to `N+M-1` are left empty in this example?